### PR TITLE
Analyzer - Add support to store values of metrics in data diagnosis

### PR DESCRIPTION
**Pull request by yukirora**
**Description**
Add support to store values of metrics in data diagnosis.

Take the following rules as example: 
```
    nccl_store_rule:
      categories: NCCL_DIS
      store: True
      metrics:
        - nccl-bw:allreduce-run0/allreduce_1073741824_busbw
        - nccl-bw:allreduce-run1/allreduce_1073741824_busbw
        - nccl-bw:allreduce-run2/allreduce_1073741824_busbw
        - nccl-bw:allreduce-run3/allreduce_1073741824_busbw
        - nccl-bw:allreduce-run4/allreduce_1073741824_busbw
    nccl_rule:
      function: multi_rules
      criteria: 'lambda label:True if min(label["nccl_store_rule"].values())/max(label["nccl_store_rule"].values())<0.95 else False'
      categories: NCCL_DIS
```
**nccl_store_rule** will store the values of the metrics in dict and save them into `label["nccl_store_rule"]` , and then **rccl_rule** can use the values of metrics through `label["nccl_store_rule"].values()` in criteria

### CODE REVIEW

**Comment by cp5555** [comment]
Would you please give an example in description for this feature?

File: tests/analyzer/test_data_diagnosis.py [hunk 1]
```diff
@@ -356,9 +360,57 @@ def test_mutli_rules(self):
...
+                'rules': {
+                    'rule1': {
+                        'categories':
+                        'NCCL_DIS',
+                        'criteria':
+                        'lambda x:x==0',
+                        'function':
+                        'value',
+                        'store':
+                        True,
```
Comment (cp5555) [inline tests/analyzer/test_data_diagnosis.py:380]: Strange format

File: tests/analyzer/test_data_diagnosis.py [hunk 2]
```diff
@@ -356,9 +360,57 @@ def test_mutli_rules(self):
...
+        diag1 = DataDiagnosis()
+        # test _run_diagnosis_rules_for_single_node
+        rules = {
+            'superbench': {
+                'rules': {
+                    'rule1': {
+                        'categories':
+                        'NCCL_DIS',
+                        'criteria':
+                        'lambda x:x==0',
```
Comment (cp5555) [inline tests/analyzer/test_data_diagnosis.py:376]: Do we need to write criteria for this 'store'?

File: tests/analyzer/test_data_diagnosis.py [hunk 3]
```diff
@@ -332,9 +333,12 @@ def test_mutli_rules(self):
...
                     'rule3': {
-                        'categories': 'CNN',
-                        'function': 'multi_rules',
-                        'criteria': 'lambda label:True if label["rule1"]+label["rule2"]>=2 else False'
+                        'categories':
+                        'CNN',
+                        'function':
+                        'multi_rules',
+                        'criteria':
+                        'lambda label:True if label["rule1"]["violated"]+label["rule2"]["violated"]>=2 else False'
```
Comment (cp5555) [inline tests/analyzer/test_data_diagnosis.py:341]: Why use this format?

File: superbench/analyzer/data_diagnosis.py [hunk 4]
```diff
@@ -169,7 +169,11 @@ def _run_diagnosis_rules_for_single_node(self, node):
                 violated_num = rule_op(data_row, self._sb_rules[rule], summary_data_row, details, categories)
             # label the node as defective one
             if self._sb_rules[rule]['store']:
-                violation[rule] = violated_num
+                store_values[rule] = {}
+                store_values[rule]['metrics'] = {}
+                for metric in self._sb_rules[rule]['metrics']:
+                    store_values[rule]['metrics'][metric] = data_row[metric]
+                store_values[rule]['violated'] = violated_num
```
Comment (cp5555) [inline superbench/analyzer/data_diagnosis.py:176]: Big concern to change original features. It mean that the file is not back-compitable. Why do we support in this way?
I think we should change the `store` part and `label["rule1"]` still store the violation results if you have criteria

File: tests/analyzer/test_data_diagnosis.py (hunk 1 above)
Comment (yukirora) [reply tests/analyzer/test_data_diagnosis.py:373]: this is by auto-format, I can not control it

File: superbench/analyzer/data_diagnosis.py (hunk 4 above)
Comment (yukirora) [reply superbench/analyzer/data_diagnosis.py]: do you mean if store is True and 'criteria' is None, we store the values of the metrics, elif store is Ture and 'criteria' is not None, we still store the violation results of the rule?

**Review by abuccts — APPROVED** [review]

**Review by cp5555 — APPROVED** [review]

File: tests/analyzer/test_data_diagnosis.py (hunk 3 above)
Comment (yukirora) [reply tests/analyzer/test_data_diagnosis.py]: I tried Yifan's suggestions, but nothing changed

**Pull request merged**